operator/README.md

```diff
@@ -0,0 +1,150 @@
+# Teleport Kubernetes Operator
+
+This package implements [an operator for Kubernetes](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).
+This operator is useful to manage Teleport resources e.g. users and roles from [Kubernetes custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+
+For more details, read the corresponding [RFD](https://github.com/gravitational/teleport-plugins/blob/master/rfd/0001-kubernetes-manager.md).
+
+## Architecture
+Teleport Operator is a K8S operator based on `operator-sdk`.
+The operator lives right beside the Teleport server, using a sidecar approach.
+
+### Setup
+When the operator starts it:
+- grabs the leader lock, to ensure only one operator is acting upon the modifications
+- ensures the exclusive role and user exist
+- creates an identity file using that user
+- starts a new client, using that identity file and manages the resources
+- install the CRDs for the managed resources
+
+After this point, the operator will listen to modifications in the K8S cluster, and act upon
+
+### Reconciliation
+When something changes (either by a `kubectl apply` or from any other source), we start the reconciliation.
+
+First, we try to identify the operation type: deletion or creation/modification.
+
+If it's a deletion and we have our own finalizer, we remove the object in Teleport and remove the finalizer.
+K8S will auto-remove the object when it gets 0 finalizers.
+
+If it's a creation/modification:
+- we add the finalizer, if it's not there already
+- we lookup the object in Teleport side, and if it's not there, we set the Origin label to `kubernetes`
+- we either create or update the resource in Teleport
+
+The first two steps above may change the object's state in K8S. If they do, we update and finish the cycle - we'll receive another reconciliation request with the new state.
+
+### Diagram
```

Comment:
I'd probably label this diagram with "Pod" above the larger box, to make it clearer that the operator runs as a sidecar container in the teleport pod.

I'd maybe also clarify "k8s" as "kube-apiserver".

It's been a while since I've worked with Kubernetes controllers, but as far as I can remember, "k8s" doesn't really maintain a "reconciler loop" as such. It's entirely up to the operator/controller to configure watches/caches against the API and trigger reconciliations based on that.

Reply:
Updated accordingly
I removed the `reconciler loop` box, I think it's ok to label the entire system as a single box and only expose the necessary details regarding interactions it has from the user's and developer's point of view.